lib/march_hare/session.rb

```diff
@@ -329,6 +332,7 @@ def automatically_recover
       new_connection = converting_rjc_exceptions_to_ruby do
         reconnecting_on_network_failures(ms) { build_new_connection }
       end
+      return if @session_closed
```

Comment:
Should we return *before* trying to reconnect?

Reply:
ok so line 324 is I think only really needed inside the test because there after `close` `automatically_recover` is called explicitly (like here https://github.com/ruby-amqp/march_hare/blob/master/spec/higher_level_api/integration/connection_spec.rb#L62-L65 ) and I didn't want to change the specs

I think in real after a call to close people should probably just create a new connection.

This return here is needed so after the rescue/retry loop is stopped it would drop back in here and try to call other hooks (and write to already closed log devices)

In my case I had the process usually stuck inside `reconnecting_on_network_failures`. Hope this makes sense.

Reply:
Log devices?

Instead of returning early from a method that's not supposed to do anything if a connection has been already closed (we can even consider throwing an exception in that case) we first try to open a connection and then return. This doesn't make any sense. Either return or thrown an exception or perhaps reconnect. This code does something in between.

Reply:
Instead of complicating connection state, have you considered making this new ivar settable and resetting it in the test in question? It would never be obvious to anyone reading the code why it was introduced. Again, if this state is supposed to be illegal, `Bunny::Session#automatic_recovery` should probably throw an `AlreadyClosedException or similar. Or work as it does today since making the retry in `reconnecting_on_network_failures` should be sufficient?

Reply:
If you pass a logger instance on `Session#connect` I had in some cases (e.g. on shutdown) that log device was already closed while this recovery part hasn't finished yet and wanted to still write to the logger.
Line 335 simply makes sure to exit the `automatically_recover` method silently after returning from `reconnecting_on_network_failures` in case a `Session#close` happened. This is not a big deal just prevents some ugly log messages.

The reason I just return here and don't raise and error (or reraise the one from `reconnecting_on_network_failures`) is because we explicitly wanted to close the connection and I'm not interested in an error telling me it couldn't reconnect. IMHO that would be a weird error to receive after calling `close`.

At the minimum IMHO we need some way to exit the rescue/retry loop inside `reconnecting_on_network_failures`. My idea was to track if `close` was called explicitly and exit the rescue/retry loop (lines 183, 229 & 594). With that my exploratory testing worked and the process was no longer stuck in the rescue/retry loop.

Specs failed with that so I added 324 but this is certainly not understandable from the code alone. I don't really want to make this internal variable directly settable from the outside but one option could be to simply use `instance_variable_set` for the few tests that want to reconnect after a explicit close without creating a new session instance. I guess the question here is if `Session#automatic_recovery` should be part of the public api and should be called explicitly in the first place.


Hope I managed to explain my reasoning a bit better.


Bunny seems to do this a bit different. If I understand the code correct this part handles the retry during the network reconnect ( I assume `recover_from_network_failure` in bunny is roughly the equivalent to `reconnecting_on_network_failures` in march_hare)
https://github.com/ruby-amqp/bunny/blob/master/lib/bunny/session.rb#L771-L784
This appears to have a limit to how often it retries via this counter https://github.com/ruby-amqp/bunny/blob/7fb05abf36637557f75a69790be78f9cc1cea807/lib/bunny/session.rb#L793 But I didn't dig to deep into that code.